File: scripts/done-means/680_quarantine_loud_driver.py
"""DONE-MEANS driver (client half) for #680 — a quarantine drop is never silent.

Drives the REAL ``JsonlSpool`` through five genuine consecutive replay failures
(the shipped ``DEFAULT_QUARANTINE_THRESHOLD``) and then asks the question the
pre-flight doc asks: after the drop, does any operator-facing count still know
the records exist?

WHY THIS SHAPE. The defect is not "quarantine happens" — quarantine is designed
(``docs/memory-limits.md:25``). The defect is that the capture lane's own
observability reads the LIVE spool only (``outage.spool_pending``), so the
quarantine move makes the number go DOWN: 15 real turns became pending=0 on
2026-07-30 and nothing said a word. So every clause here is about what a READER
sees after the drop, never about whether the sidecar file was written.

Content-free output: clause names, states, and counts only — never payloads.
"""

from __future__ import annotations

import json
import sys
import tempfile
from pathlib import Path
from typing import Any

REPO_ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(REPO_ROOT / "python" / "openbrain-memory" / "src"))
sys.path.insert(0, str(REPO_ROOT / "python" / "openbrain" / "src"))

from openbrain_memory.spool import (  # noqa: E402
    DEFAULT_QUARANTINE_THRESHOLD,
    JsonlSpool,
)

FAILURES: list[str] = []


def check(name: str, ok: bool, detail: str) -> None:
    print(f"{'PASS' if ok else 'FAIL'}  {name}: {detail}")
    if not ok:
        FAILURES.append(name)


_MISSING = object()


def resolve(module: Any, name: str) -> Any:
    """Look up a symbol that may not exist yet, WITHOUT crashing the run.

    A bare ``module.new_function(...)`` raises ``AttributeError`` at the
    pre-fix tree and kills every clause after it — a crash upstream of the
    measurement, which is a FALSE RED indistinguishable in shape from a real
    one and reached by the ordinary act of checking for a capability that does
    not exist yet (``docs/lane-contract.md`` round 18, the Python spelling of
    the dynamic-import rule). Resolving through a sentinel lets each clause
    report its own honest FAIL for the defect's own reason.
    """
    return getattr(module, name, _MISSING)


def call(fn: Any, *args: Any) -> Any:
    """Call a resolved symbol, or return the sentinel when it is absent."""
    if fn is _MISSING:
        return _MISSING
    return fn(*args)


def quarantine_one_unit(spool_path: Path) -> JsonlSpool:
    """Append one unit and fail its replay exactly ``threshold`` times.

    Returns the spool, with the unit quarantined by the real code path — no
    hand-written sidecar. A hand-built fixture would prove the fixture.
    """
    spool = JsonlSpool(spool_path)

    spool.append(
        operation="ingest_raw_turn",
        payload={"turns": [{"turn_uuid": "done-means-680-turn", "role": "user"}]},
    )

    def always_fails(record: Any) -> Any:
        raise RuntimeError("done-means forced delivery failure")

    for _ in range(DEFAULT_QUARANTINE_THRESHOLD):
        spool.replay_with_report(always_fails)

    return spool


def main() -> int:
    print(f"INFO  quarantine threshold under test: {DEFAULT_QUARANTINE_THRESHOLD}")

    with tempfile.TemporaryDirectory(prefix="done-means-680-") as tmp:
        spool_path = Path(tmp) / "claude-spool.jsonl"
        spool = quarantine_one_unit(spool_path)
        status = spool.status()

        # ------------------------------------------------------------------
        # (setup) The precondition every other clause reasons about: five
        # forced failures really did quarantine the unit and empty the live
        # spool. If this is not true the rest of the run proves nothing, so it
        # is asserted rather than assumed.
        # ------------------------------------------------------------------
        check(
            "setup-quarantined",
            status.quarantined_count == 1 and status.pending_count == 0,
            f"{DEFAULT_QUARANTINE_THRESHOLD} forced failures -> "
            f"quarantined_count={status.quarantined_count} "
            f"pending_count={status.pending_count} "
            "(the live spool is empty and the records are in the sidecar)",
        )

        # ------------------------------------------------------------------
        # (a) THE DEFECT ITSELF. The capture lane's operator-facing count must
        # not read 0 while records sit abandoned in the sidecar. This is the
        # exact live symptom: /health said spool_pending:0 with 15 turns gone.
        #
        # Asserts the COUNT, not the presence of a function: a reader that
        # returns 0 here is the shipped defect regardless of what it is named.
        # ------------------------------------------------------------------
        from openbrain.apps.capture import outage  # noqa: PLC0415

        pending = outage.spool_pending(spool_path)
        check(
            "a-pending-not-blind",
            pending is not None and pending > 0,
            f"after the drop, the capture lane's pending count reads {pending!r} "
            "(0 or None means the records are invisible to every operator surface)",
        )

        # ------------------------------------------------------------------
        # (b) LOUD, AND SPECIFICALLY ABOUT QUARANTINE. A count that merely
        # went up could be an ordinary outage backlog — indistinguishable from
        # the healthy "held for replay" case an operator is trained to ignore.
        # The notice must NAME the abandoned records, because the two states
        # need different operator actions: an outage backlog drains itself, a
        # quarantined unit never will.
        #
        # Both halves in ONE clause (round 17): a count that rose AND a notice
        # that says the word. Split, each half passes for the wrong reason.
        # ------------------------------------------------------------------
        spool_quarantined = resolve(outage, "spool_quarantined")
        quarantined = call(spool_quarantined, spool_path)
        notice = (
            outage.spool_notice(outage.DEGRADED_NOTICE, pending, quarantined)
            if quarantined is not _MISSING
            else "<no quarantine reader exists>"
        )
        check(
            "b-notice-names-quarantine",
            quarantined == 1 and "quarantin" in notice.lower(),
            f"quarantined={quarantined!r} and the operator notice names it: {notice!r}",
        )

        # ------------------------------------------------------------------
        # (c) CONTROL — HEALTHY STAYS QUIET (round 8: every failure-signal
        # check needs a control proving the healthy path stays healthy).
        # A spool with nothing quarantined must produce the ORDINARY notice
        # with no quarantine wording. Without this, "always shout quarantine"
        # passes (a) and (b) and trains the operator to ignore the line.
        # ------------------------------------------------------------------
        clean_path = Path(tmp) / "clean-spool.jsonl"
        clean = JsonlSpool(clean_path)
        clean.append(
            operation="ingest_raw_turn",
            payload={"turns": [{"turn_uuid": "done-means-680-healthy", "role": "user"}]},
        )
        clean_pending = outage.spool_pending(clean_path)
        clean_quarantined = call(spool_quarantined, clean_path)
        clean_notice = (
            outage.spool_notice(
                outage.DEGRADED_NOTICE, clean_pending, clean_quarantined
            )
            if clean_quarantined is not _MISSING
            else "<no quarantine reader exists>"
        )
        check(
            "c-control-healthy-quiet",
            clean_quarantined == 0
            and clean_pending == 1
            and "quarantin" not in clean_notice.lower(),
            f"an ordinary held turn reports quarantined={clean_quarantined} "
            f"pending={clean_pending} and says nothing about quarantine: {clean_notice!r}",
        )

        # ------------------------------------------------------------------
        # (d) UNREADABLE IS NOT ZERO. `spool_pending` already distinguishes
        # "absent -> 0" from "unreadable -> None" on purpose, because a count
        # is the one thing on that line an operator acts on. The quarantine
        # reader inherits that contract: guessing 0 for a sidecar it could not
        # read would recreate this very defect one directory over.
        #
        # Driven by pointing the reader at a DIRECTORY where the sidecar file
        # is expected — a real OSError from the real code path, not a mock.
        # ------------------------------------------------------------------
        unreadable_spool = Path(tmp) / "unreadable-spool.jsonl"
        unreadable_spool.write_text("", encoding="utf-8")
        sidecar_dir = Path(str(unreadable_spool) + ".quarantine.jsonl")
        sidecar_dir.mkdir()
        unreadable = call(spool_quarantined, unreadable_spool)
        check(
            "d-unreadable-is-none",
            unreadable is None,
            f"an unreadable quarantine sidecar reads {unreadable!r}, not a guessed 0",
        )

        # ------------------------------------------------------------------
        # (e) ABSENT IS GENUINELY ZERO. The other side of (d): no sidecar at
        # all is a real, knowable zero and must NOT degrade to None, or every
        # healthy machine on earth prints an unknown-count line forever.
        # ------------------------------------------------------------------
        absent = call(spool_quarantined, Path(tmp) / "never-existed.jsonl")
        check(
            "e-absent-is-zero",
            absent == 0,
            f"a spool with no quarantine sidecar reads {absent!r}",
        )

        # ------------------------------------------------------------------
        # (f) THE SIDECAR IS INTACT AND REPLAYABLE. "Never a silent drop" is
        # only half the promise; the records must still BE there for the
        # operator's replay decision (issue #680 item 4). Reads the envelope
        # back and asserts the abandoned unit is identifiable by operation and
        # spool key — content-free, no payload assertions.
        # ------------------------------------------------------------------
        sidecar = Path(str(spool_path) + ".quarantine.jsonl")
        envelopes = [
            json.loads(line)
            for line in sidecar.read_text(encoding="utf-8").splitlines()
            if line.strip().startswith("{") and '"schema"' in line
        ]
        check(
            "f-sidecar-replayable",
            len(envelopes) == 1
            and envelopes[0].get("consecutive_failures") == DEFAULT_QUARANTINE_THRESHOLD,
            f"the sidecar holds {len(envelopes)} envelope(s) recording "
            f"consecutive_failures="
            f"{envelopes[0].get('consecutive_failures') if envelopes else 'n/a'} "
            "— the records survive for an operator replay decision",
        )

    print()
    if FAILURES:
        print(f"CLIENT-HALF FAIL — {len(FAILURES)} clause(s): {', '.join(FAILURES)}")
        return 1
    print("CLIENT-HALF PASS — a quarantine drop is visible and named at the capture lane.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/done-means/680-quarantine-loud.driver.ts
/**
 * DONE-MEANS driver (server half) for #680 — a quarantined unit raises a LOUD
 * fault in `/health`, and can never be reported as a healthy empty spool.
 *
 * ---------------------------------------------------------------------------
 * THE DEFECT, at this boundary
 * ---------------------------------------------------------------------------
 * `server/capture/liveness-observer.ts:264-265` hardcodes `spoolPending: 0,
 * outageAnnouncements: 0`, and `TransportCaptureHealth`
 * (`server/transport/health.ts:54-78`) has no quarantine field at all. So on
 * 2026-07-30, with 15 real turns abandoned in a sidecar, the live `/health`
 * read `spool_pending:0, reason:"capture lane delivering"` — a green verdict
 * published over a permanent data loss.
 *
 * The module's own docstring argues, correctly for `spool_pending`, that a
 * SERVER cannot enumerate client-side spool files. That argument does NOT
 * extend to quarantine: a quarantine count is not something the server must
 * discover by enumeration, it is something a client REPORTS, exactly as the
 * capture lane already reports its other counts into this observation. The
 * delta is therefore a reported input, not an invented server-side reader.
 *
 * ---------------------------------------------------------------------------
 * SUBJECT: the real judge and the real composition
 * ---------------------------------------------------------------------------
 * Clauses drive `readCaptureLiveness` — the single judge of record, the same
 * function `server/application/index.ts:204` calls — and the verdict clauses
 * bind it to a live HTTP `/health` through the real `createShadowApplication`
 * composition on an ephemeral 127.0.0.1 port. A check that rebuilds its
 * subject proves the rebuild (#624 harvest).
 *
 * Imports of the NEW field/reader are DYNAMIC. A static import of something
 * that does not exist at the pre-fix tree dies at module resolution before any
 * clause prints — a false RED indistinguishable in shape from a real one, and
 * reached by the ordinary act of writing a check for a capability that does
 * not exist yet (docs/lane-contract.md round 18).
 *
 * No database. No network beyond a kernel-assigned ephemeral port. No
 * wall-clock verdict (round 5, #632/#634).
 */
import express from "express";
import { Writable } from "node:stream";
import pino from "pino";
import type { Logger } from "pino";
import type { AddressInfo } from "node:net";
import type { Server } from "node:http";
import { createShadowApplication } from "../../server/application/index.ts";
import type { ShadowApplication } from "../../server/application/index.ts";
import { parseServerConfig } from "../../server/config.ts";
import type { ServerConfig } from "../../server/config.ts";
import type { Database, DatabaseHealth } from "../../server/db/pool.ts";

const results: Array<{ clause: string; ok: boolean }> = [];

function clause(name: string, ok: boolean, detail: string): void {
  results.push({ clause: name, ok });
  console.log(`${ok ? "PASS" : "FAIL"}  (${name}) ${detail}`);
}

const CONNECTED: DatabaseHealth = { connected: true, total: 2, idle: 2, waiting: 0 };

function testConfig(): ServerConfig {
  const result = parseServerConfig({
    DB_HOST: "db.internal",
    DB_NAME: "open_brain_test",
    DB_USER: "open_brain",
    LOG_FILE: "logs/open-brain.log",
    OPEN_BRAIN_SERVER_IP: "192.0.2.21",
  });
  if (!result.ok) {
    throw new Error(`invalid driver configuration: ${JSON.stringify(result.issues)}`);
  }
  return result.config;
}

function fakeDatabase(): Database {
  return {
    pool: {} as Database["pool"],
    close: async () => {},
    health: async () => CONNECTED,
  } as Database;
}

function capturingLogger(): { logger: Logger; lines: () => string[] } {
  const captured: string[] = [];
  const stream = new Writable({
    write(chunk: Buffer, _encoding, next) {
      captured.push(chunk.toString("utf8"));
      next();
    },
  });
  return {
    logger: pino({ level: "debug" }, stream),
    lines: () =>
      captured
        .join("")
        .split("\n")
        .filter((line) => line.trim().length > 0),
  };
}

interface LiveApp {
  readonly base: string;
  close(): Promise<void>;
}

async function listen(
  overrides: Record<string, unknown>,
  logger: Logger,
): Promise<LiveApp> {
  const application: ShadowApplication = createShadowApplication({
    config: testConfig(),
    logger,
    database: fakeDatabase(),
    authenticate: ((
      _request: unknown,
      _response: unknown,
      next: () => void,
    ) => next()) as never,
    parseRequestBody: express.json(),
    serverFactory: () => ({ connect: async () => {} }) as never,
    fetch: (async () => new Response("{}", { status: 200 })) as unknown as typeof fetch,
    ...(overrides as object),
  } as never);

  const server: Server = await new Promise((resolve) => {
    const listener = application.app.listen(0, "127.0.0.1", () => resolve(listener));
  });
  const { port } = server.address() as AddressInfo;
  return {
    base: `http://127.0.0.1:${port}`,
    close: async () => {
      await new Promise<void>((resolve) => server.close(() => resolve()));
      await application.close();
    },
  };
}

interface HealthBody {
  readonly capture?: Record<string, unknown>;
  readonly [key: string]: unknown;
}

async function getHealth(
  base: string,
): Promise<{ status: number; body: HealthBody }> {
  const response = await fetch(`${base}/health`);
  return {
    status: response.status,
    body: (await response.json()) as HealthBody,
  };
}

/**
 * A healthy delivering observation — the CONTROL baseline every fault clause
 * mutates one field of. Built as a plain object so a missing quarantine field
 * at the pre-fix tree is a value-level absence, never an import failure.
 */
function deliveringObservation(): Record<string, unknown> {
  return {
    sessionsObserved: 4,
    watermarkBytesAdvanced: 120,
    spoolPending: 0,
    outageAnnouncements: 0,
    turnsByRole: { user: 60, assistant: 60 },
    silenceSeconds: 5,
  };
}

async function main(): Promise<number> {
  const observer = await import("../../server/capture/liveness-observer.ts");
  const readCaptureLiveness = observer.readCaptureLiveness as (
    o: Record<string, unknown> | undefined,
  ) => Record<string, unknown> | undefined;

  // --------------------------------------------------------------------
  // (a) THE JUDGE SEES QUARANTINE AT ALL. A quarantined unit reported by
  // the capture lane must make the reading STALE and must be named in the
  // reason. At the pre-fix tree the field does not exist in the shape, so
  // the judge ignores it entirely and returns "capture lane delivering" —
  // the exact live 2026-07-30 verdict published over 15 lost turns.
  // --------------------------------------------------------------------
  const quarantinedReading = readCaptureLiveness({
    ...deliveringObservation(),
    spoolQuarantined: 3,
  });
  const staleOnQuarantine = quarantinedReading?.stale === true;
  const reason = String(quarantinedReading?.reason ?? "");
  clause(
    "a-quarantine-is-stale",
    staleOnQuarantine && /quarantin/i.test(reason),
    `an observation carrying 3 quarantined unit(s) reads stale=${String(
      quarantinedReading?.stale,
    )} reason=${JSON.stringify(reason)}`,
  );

  // --------------------------------------------------------------------
  // (b) THE COUNT IS PUBLISHED, not merely folded into a boolean. An
  // operator deciding replay-vs-accept needs the number; a bare "stale"
  // flag makes them go and read the sidecar to learn the size of the loss.
  // Asserts the exact value survives to the block, so an implementation
  // that clamps or booleans it fails.
  // --------------------------------------------------------------------
  clause(
    "b-count-published",
    quarantinedReading?.quarantined_count === 3,
    `the block publishes quarantined_count=${JSON.stringify(
      quarantinedReading?.quarantined_count,
    )} (expected the reported 3)`,
  );

  // --------------------------------------------------------------------
  // (c) CONTROL — a delivering lane with NOTHING quarantined stays GREEN
  // and still publishes the field as a real zero. Fails any implementation
  // that shouts on every reading, and any that omits the field when it is
  // zero (an absent field is unreadable to a monitor that must alert on
  // it crossing 0 -> 1). PASSES PRE-FIX for the verdict half by design:
  // a check that fails everywhere proves only that it fails (round 13).
  // --------------------------------------------------------------------
  const healthyReading = readCaptureLiveness({
    ...deliveringObservation(),
    spoolQuarantined: 0,
  });
  clause(
    "c-control-healthy-green",
    healthyReading?.stale === false &&
      healthyReading?.quarantined_count === 0 &&
      !/quarantin/i.test(String(healthyReading?.reason ?? "")),
    `a delivering lane with nothing quarantined reads stale=${String(
      healthyReading?.stale,
    )} quarantined_count=${JSON.stringify(
      healthyReading?.quarantined_count,
    )} reason=${JSON.stringify(String(healthyReading?.reason ?? ""))}`,
  );

  // --------------------------------------------------------------------
  // (d) UNOBSERVABLE IS ANNOUNCED, NOT ZEROED. The module already
  // publishes `observableFaults` so a green block is not misread as "all
  // faults checked and clear". A vantage point with no quarantine input
  // must NOT report a confident 0 — that is precisely the shape of the
  // defect (a hardcoded 0 standing in for an unmeasured quantity). Absent
  // input reports the field as absent/undefined, never 0.
  // --------------------------------------------------------------------
  const unreportedReading = readCaptureLiveness(deliveringObservation());
  clause(
    "d-unobserved-is-not-zero",
    unreportedReading?.stale === false &&
      unreportedReading?.quarantined_count === undefined,
    `an observation that reports NO quarantine input publishes quarantined_count=${JSON.stringify(
      unreportedReading?.quarantined_count,
    )} (absent, not a fabricated 0)`,
  );

  // --------------------------------------------------------------------
  // (e) IT REACHES A LIVE /health, DEGRADED. The judge being right is not
  // the promise; the promise is that an operator watching the endpoint
  // sees it. Drives the REAL composition over real HTTP and asserts both
  // the transport status and the published block.
  // --------------------------------------------------------------------
  const { logger } = capturingLogger();
  const app = await listen(
    {
      captureObserver: () => ({
        ...deliveringObservation(),
        spoolQuarantined: 3,
      }),
    },
    logger,
  );
  try {
    const health = await getHealth(app.base);
    const capture = health.body.capture ?? {};
    clause(
      "e-live-health-degraded",
      health.status === 503 &&
        capture.stale === true &&
        capture.quarantined_count === 3 &&
        /quarantin/i.test(String(capture.reason ?? "")),
      `live GET /health -> ${health.status} with capture.stale=${String(
        capture.stale,
      )} capture.quarantined_count=${JSON.stringify(
        capture.quarantined_count,
      )} reason=${JSON.stringify(String(capture.reason ?? ""))}`,
    );
  } finally {
    await app.close();
  }

  // --------------------------------------------------------------------
  // (f) CONTROL — ABSENCE IS NOT STALENESS survives the change. A
  // deployment composing no capture observer publishes no block and stays
  // 200/green. PASSES PRE-FIX by design (rounds 8 and 13): the point is
  // that this fix does not make an opted-out worker degrade itself.
  // --------------------------------------------------------------------
  const { logger: bareLogger } = capturingLogger();
  const bare = await listen({}, bareLogger);
  try {
    const health = await getHealth(bare.base);
    clause(
      "f-control-absence-not-staleness",
      health.status === 200 && health.body.capture === undefined,
      `a deployment composing no observer -> ${health.status} with no capture block`,
    );
  } finally {
    await bare.close();
  }

  const failed = results.filter((r) => !r.ok);
  console.log();
  if (failed.length > 0) {
    console.log(
      `SERVER-HALF FAIL — ${failed.length} clause(s): ${failed
        .map((r) => r.clause)
        .join(", ")}`,
    );
    return 1;
  }
  console.log("SERVER-HALF PASS — a quarantined unit is a loud, counted fault in /health.");
  return 0;
}

main()
  .then((code) => {
    process.exit(code);
  })
  .catch((error: unknown) => {
    // A top-level await with no .catch exits 0 when it throws — a crashing
    // subject banking a false GREEN (docs/lane-contract.md round 13).
    console.error(
      `SERVER-HALF ERROR — driver crashed: ${
        error instanceof Error ? `${error.name}: ${error.message}` : String(error)
      }`,
    );
    process.exit(1);
  });
